### PR TITLE
feat(core): make rerun work with new storage

### LINKED ISSUE #2258: implement renku rerun with new database
`renku rerun` currently doesn't work (it still uses old graph building or provenance graph). We have to implement it again with the new Database and ActivityGateway. 

Reenable the `renku rerun` tests to verify this works as before.

### PR DESCRIPTION
**Pull request by mohammad-sdsc**
# Description

Reimplementation of the `renku run <paths>` using the new persistent layer. Support `--from` flag to start execution from a specific path. Returns 1 if no workflow is re-executed.
Followup story: https://github.com/SwissDataScienceCenter/renku-python/issues/2323

Fixes #2258

### CODE REVIEW

File: renku/core/management/interface/activity_gateway.py [hunk 1]
```diff
@@ -35,13 +36,19 @@ def get_plans_and_usages_for_latest_activities(self) -> Dict[AbstractPlan, List[
         """Get all usages associated with a plan by its latest activity."""
         raise NotImplementedError
 
+    def get_activities_by_generation(self, path: Union[Path, str]) -> List[Activity]:
+        """Return the list of all activities that generate a path."""
+
```
Comment (Panaetius) [inline renku/core/management/interface/activity_gateway.py:41]: Should raise NotImplementedException

File: renku/core/management/interface/activity_gateway.py [hunk 2]
```diff
@@ -35,13 +36,19 @@ def get_plans_and_usages_for_latest_activities(self) -> Dict[AbstractPlan, List[
...
     def get_downstream_activities(self, activity: Activity, max_depth=None) -> Set[Activity]:
         """Get downstream activities that depend on this activity."""
         raise NotImplementedError
 
     def get_downstream_activity_chains(self, activity: Activity) -> List[Tuple[Activity, ...]]:
         """Get a list of tuples of all downstream paths of this activity."""
 
+    def get_upstream_activity_chains(self, activity: Activity) -> List[Tuple[Activity, ...]]:
+        """Get a list of tuples of all upstream paths of this activity."""
+
```
Comment (Panaetius) [inline renku/core/management/interface/activity_gateway.py:51]: Same as above

**Review by Panaetius — CHANGES_REQUESTED** [review]

File: renku/core/commands/rerun.py [hunk 3]
```diff
@@ -35,23 +41,70 @@ def rerun_workflows():
...
+
+    plans = [a.plan_with_values for a in activities]
+
+    execute_workflow(plans=plans, command_name="rerun")
+
+
+def _get_activities(paths: List[str], sources: List[str], activity_gateway: IActivityGateway):
+    all_activities = defaultdict(set)
+
+    def include_newest_activity(activity):
```
[the remark this answers is not in the document]
Comment (mohammad-sdsc) [reply renku/core/commands/rerun.py:67]: I've added a more general utility function and used it here. There is still a duplicated line but I believe it's more readable and re-usable this way.

**Review by Panaetius — DISMISSED** [review]
Thank you!

**Review by Panaetius — APPROVED** [review]

**Pull request merged**